lib/Controller/BookingController.php

```diff
@@ -86,60 +113,143 @@ public function getBookableSlots(int $appointmentConfigId,
-			->setTime(0, 0)
-			->getTimestamp();
-		$endTimeInTz = (new DateTimeImmutable())
-			->setTimestamp($end)
-			->setTimezone($tz)
-			->setTime(23, 59, 59)
-			->getTimestamp();
-
-		if ($startTimeInTz > $endTimeInTz) {
+							 string $description,
+							 string $timeZone): JsonResponse {
+		if ($start > $end) {
 			return JsonResponse::fail('Invalid time range', Http::STATUS_UNPROCESSABLE_ENTITY);
 		}
 
 		try {
 			$config = $this->appointmentConfigService->findById($appointmentConfigId);
 		} catch (ServiceException $e) {
+			$this->logger->error('No appointment config found for id ' . $appointmentConfigId, ['exception' => $e]);
+			return JsonResponse::fail(null, Http::STATUS_NOT_FOUND);
+		}
+		try {
+			$booking = $this->bookingService->book($config, $start, $end, $timeZone, $displayName, $email, $description);
+		} catch (ClientException $e) {
+			$this->logger->warning('No slot available for start: ' . $start . ', end: ' . $end . ', config id: ' . $appointmentConfigId , ['exception' => $e]);
 			return JsonResponse::fail(null, Http::STATUS_NOT_FOUND);
+		} catch (InvalidArgumentException $e) {
+			$this->logger->warning($e->getMessage(), ['exception' => $e]);
+			return JsonResponse::fail(null, Http::STATUS_UNPROCESSABLE_ENTITY);
+		} catch (ServiceException $e) {
+			$this->logger->error($e->getMessage(), ['exception' => $e]);
+			return JsonResponse::errorFromThrowable($e, $e->getHttpCode());
+		}
+
+		return JsonResponse::success($booking);
+	}
+
+	/**
+	 * @PublicPage
+	 * @NoCSRFRequired
+	 *
+	 * @param string $token
+	 * @return TemplateResponse
+	 * @throws Exception
+	 */
+	public function confirmBooking(string $token): TemplateResponse {
+		try {
+			$booking = $this->bookingService->findByToken($token);
+		} catch(ClientException $e) {
+			$this->logger->warning($e->getMessage(), ['exception' => $e]);
+			return new TemplateResponse(
+				Application::APP_ID,
+				'appointments/404-booking',
+				[],
+				TemplateResponse::RENDER_AS_GUEST
+			);
 		}
 
-		// validate slot is available
-		// write slot, preparation duration and follow up duration events
-//		$this->bookingService->book($startTimeInTz, $endTimeInTz, $config, $start);
+		try {
+			// This should never happen
+			$config = $this->appointmentConfigService->findById($booking->getApptConfigId());
+		} catch (ServiceException $e) {
+			$this->logger->error($e->getMessage(), ['exception' => $e]);
+			return new TemplateResponse(
+				Application::APP_ID,
+				'appointments/404-booking',
+				[],
+				TemplateResponse::RENDER_AS_GUEST
+			);
+		}
 
-		return JsonResponse::success();
+		try {
+			// This happens most likely bc of conflicting slots
+			$booking = $this->bookingService->confirmBooking($booking, $config);
+		} catch (ClientException $e) {
+			$this->logger->warning($e->getMessage(), ['exception' => $e]);
+			$link = $this->urlGenerator->linkToRouteAbsolute( 'calendar.booking.getBookableSlots', [ 'appointmentConfigId' => $config->getId() ]);
```

Comment:
```suggestion
			$link = $this->urlGenerator->linkToRouteAbsolute( 'calendar.appointment.show', [ 'token' => $config->getToken() ]);
```

because you want to redirect to the *page*, not the REST API